ml_pipeline_engine/dag/manager.py

```diff
@@ -526,27 +545,46 @@ async def _run_oneof(self, node_id: NodeId) -> t.Any:
 
         logger.debug('Prepare OneOf DAG node_id=%s', node_id)
 
-        for subgraph_node_id in self.dag.graph.nodes[node_id][NodeField.oneof_nodes]:
-            logger.debug(
-                'Prepare OneOf subgraph to start target_node_id=%s, subgraph_node_id=%s',
-                node_id,
+        for idx, subgraph_node_id in enumerate(self.dag.graph.nodes[node_id][NodeField.oneof_nodes]):
+
+            oneof_dag = self._get_reduced_dag_input_one_of(
+                source=self.dag.input_node,
+                dest=subgraph_node_id,
+            )
+
+            logger.debug('Prepare [%s]%s to start. OneOf result node %s', idx, oneof_dag, node_id)
+
+            self._create_task(coro=self._run_dag(dag=oneof_dag), name=str(oneof_dag))
+
+            await self._lock_manager.wait_for_condition(
                 subgraph_node_id,
+                lambda: (
+                    self.__has_subgraph_error(oneof_dag)  # noqa: B023
+                    or self._node_storage.exists_result_type(
+                        subgraph_node_id,  # noqa: B023
+                        exclude_type=(Recurrent,),
```

Comment:
I don't get it. What is target_type that you looking for ? or you looking for any object that is not `Recurrent` 
(don't we have a specific type for result objects that here we had to look for anything in the storage) ?

Reply:
> or you looking for any object that is not Recurrent

Exactly as code says =) 

> (don't we have a specific type for result objects that here we had to look for anything in the storage) ?

Maybe, but it would add more work with artifact storage and extend the PR.